### ISSUE TITLE
False positive for UnconditionalJumpStatementInLoop when nested loop does contain condition(s)

### ISSUE THREAD

**Issue by YSakhno** [detekt/detekt#6657]
## Expected Behavior

The rule violation should not be reported in this specific (and other similar) case, see below.

## Observed Behavior

Having the code (adapted and translated to Kotlin from [here](https://docs.python.org/3/library/itertools.html#itertools.permutations); only the relevant part of the translation is shown)

```kotlin
    outer@while (true) {
        for (i in r2 - 1 downTo 0) {
            if (--cycles[i] != 0) {
                indices.swapAt(i, n - cycles[i])
                yield(indices.subList(0, r2).map { pool[it] })
                continue@outer
            }
            indices.add(indices.removeAt(i))
            cycles[i] = n - i
        }
        break
    }
```

Detekt reports a rule violation on the `break` statement:

```
Permutations.kt:28:11: This loop contains an unconditional jump expression which essentially renders it useless as it will exit the loop during the first iteration. [UnconditionalJumpStatementInLoop]
```

The code functions correctly (confirmed by numerous tests), and the outer loop _does_ perform multiple iterations, because of the `continue` statement from the nested loop.

## Steps to Reproduce

See **Observed Behavior** section above.

## Context

Rule `UnconditionalJumpStatementInLoop` is active.

## Your Environment

* Version of detekt used: **1.23.1**
* Version of Gradle used (if applicable): **8.4**
* Gradle scan link (add `--scan` option when running the gradle task): did not run; not necessary IMO
* Operating System and version: Windows 10 Pro 22H2; OS build 19045.3693
* Link to your project (if it's a public repository): not available in public repo yet

**Comment by cortinico** [comment]
Looks similar to https://github.com/detekt/detekt/issues/6442, which was released on 1.23.2

Could you try again on 1.23.2 (or even better 1.23.4) and report if this was fixed

**Comment by YSakhno** [comment]
Yep, still reports violation

![image](https://github.com/detekt/detekt/assets/153881/0b7f7bfb-48b8-4c14-af9d-8711aab7196f)

**Comment by BraisGabin** [comment]
Ok, so that fix didn't fix this new case. @YSakhno do you want (or have the time) to give it a try and send a PR?

**Comment by YSakhno** [comment]
I am not familiar with Detekt's codebase, and the logic the rule has to detect and unravel in this particular case seems to be complicated. Fixing this will probably take me much more time than someone who's well versed.

**Comment by atulgpt** [comment]
I can look into this

**Comment by atulgpt** [comment]
@YSakhno can you also post the compiling minimal snippet?

Also if your requirement to have break you can use the Kotlin guideline at `There is no direct equivalent for break, but it can be simulated by adding another nesting lambda and non-locally returning from it:` taken from https://kotlinlang.org/docs/returns.html#return-to-labels. This will remove the `while(true)`

Also will having a labelled `for` work for your case? In that case you can directly jump to for label

**Comment by YSakhno** [comment]
Here is a complete snippet that should be compilable

```kotlin
fun <E> Iterable<E>.permutations(r: Int = -1): Sequence<List<E>> = sequence {
    val pool = toList()
    val n = pool.size
    val r2 = if (r < 0) n else r

    if (r2 > n) return@sequence

    val indices = (0..<n).toMutableList()
    val cycles = (n downTo n - r2 + 1).toList().toIntArray()

    yield(indices.subList(0, r2).map { pool[it] })

    outer@while (true) {
        for (i in r2 - 1 downTo 0) {
            if (--cycles[i] != 0) {
                indices.swapAt(i, n - cycles[i])
                yield(indices.subList(0, r2).map { pool[it] })
                continue@outer
            }
            indices.add(indices.removeAt(i))
            cycles[i] = n - i
        }
        break
    }
}

private fun <E> MutableList<E>.swapAt(index1: Int, index2: Int) {
    this[index1] = this[index2].also { this[index2] = this[index1] }
}
```

While it _is_ actually possible to modify this code to remove the violation (replacing `break` with `return@sequence` is one possibility), this does not eliminate the fact that Detekt reports a false positive where it shouldn't.
